autorally_core/src/StateEstimator/StateEstimator.cpp

```diff
@@ -345,125 +416,178 @@ namespace autorally_core
           m_lastImuTgps = m_lastIMU->header.stamp.toSec();
           pre_int_data.integrateMeasurement(acc, gyro, imuDT);
           m_lastIMU = m_ImuOptQ.popBlocking();
-//          ROS_WARN("Last IMU dt was %f, put time %f", imuDT, m_lastIMU->header.stamp.toSec());
         }
-
+        // adding the integrated IMU measurements to the factor graph
         ImuFactor imuFactor(X(m_poseVelKey), V(m_poseVelKey), X(m_poseVelKey+1), V(m_poseVelKey+1), B(m_biasKey),
                   pre_int_data);
-
         newFactors.add(imuFactor);
         newFactors.add(BetweenFactor<imuBias::ConstantBias>(B(m_biasKey), B(m_biasKey+1), imuBias::ConstantBias(),
             noiseModel::Diagonal::Sigmas( sqrt(pre_int_data.deltaTij()) * noiseModelBetweenbias_sigma)));
 
-        // Predict forward to get an initialization for the pose and velocity
+
+        // Predict forward to get an initial estimate for the pose and velocity
         NavState curNavState(m_prevPose, m_prevVel);
         NavState nextNavState = pre_int_data.predict(curNavState, m_previousBias);
 
-        SharedDiagonal gpsNoise = noiseModel::Diagonal::Sigmas(Vector3(m_gpsSigma, m_gpsSigma, 3.0 * m_gpsSigma));
 
-        GPSFactor gpsFactor(G(m_poseVelKey+1), Point3(E, N, U), gpsNoise);
-        newFactors.add(gpsFactor);
+        if (usingGPS)
+        {
+          double E, N, U;
+          m_enu.Forward(fix->latitude, fix->longitude, fix->altitude, E, N, U);
 
-        BetweenFactor<Pose3> imuPgpsFactor(X(m_poseVelKey+1), G(m_poseVelKey+1), m_imuPgps,
-            noiseModel::Diagonal::Sigmas((Vector(6) << 0.001,0.001,0.001,0.03,0.03,0.03).finished()));
-        newFactors.add(imuPgpsFactor);
+          SharedDiagonal gpsNoise = noiseModel::Diagonal::Sigmas(Vector3(m_gpsSigma, m_gpsSigma, 3.0 * m_gpsSigma));
+
+          GPSFactor gpsFactor(G(m_poseVelKey+1), Point3(E, N, U), gpsNoise);
+          newFactors.add(gpsFactor);
+
+          BetweenFactor<Pose3> imuPgpsFactor(X(m_poseVelKey+1), G(m_poseVelKey+1), m_imuPgps,
+              noiseModel::Diagonal::Sigmas((Vector(6) << 0.001,0.001,0.001,0.03,0.03,0.03).finished()));
+          newFactors.add(imuPgpsFactor);
+        }
 
         newVariables.insert(X(m_poseVelKey+1), nextNavState.pose());
         newVariables.insert(V(m_poseVelKey+1), nextNavState.v());
         newVariables.insert(B(m_biasKey+1), m_previousBias);
         newVariables.insert(G(m_poseVelKey+1), nextNavState.pose().compose(m_imuPgps));
 
-        m_isam->update(newFactors, newVariables);
-        m_prevPose = m_isam->calculateEstimate<Pose3>(X(m_poseVelKey+1));
-        m_prevVel = m_isam->calculateEstimate<Vector3>(V(m_poseVelKey+1));
-        m_previousBias = m_isam->calculateEstimate<imuBias::ConstantBias>(B(m_biasKey+1));
-        //std::cout << m_isam->marginalCovariance(X(m_poseVelKey+1)) << std::endl << std::endl;
-
-        diag_ok("Still ok!");
 
+        try
         {
-          boost::mutex::scoped_lock guard(m_optimizedStateMutex);
```

Comment:
In the original, update isn't in the thread guard. I didn't remove anything from it.